muffnn/mlp/base.py

```diff
@@ -301,6 +301,26 @@ def _set_up_graph(self):
         if self._transform_layer_index == -1:
             self._transform_layer = t
 
+        # Prediction gradients (e.g., for analyzing the importance of features)
+        # We use the top layer before the output activation function
+        # (e.g., softmax, sigmoid) following
+        # https://arxiv.org/pdf/1312.6034.pdf
+        if self.is_sparse_:
+            self._prediction_gradient = None
+        else:
+            output_shape = self.output_layer_.get_shape()
+            if len(output_shape) == 1:
+                self._prediction_gradient = tf.gradients(
+                    t, self._input_values)[0]
```

Comment:
I'm a little confused by the `[0]` indexing here - why do we only want the first element?

Reply:
`tf.gradients` returns a list of gradients dy/dx, one per given x.  In this case, that has length 1.  Also, it's polymorphic, so `tf.gradients([y], [x]) == tf.gradients(y, x)`.

```
(Pdb) tf.gradients(t, [self._input_values])
[<tf.Tensor 'gradients/layer_0/MatMul_grad/MatMul:0' shape=(?, 20) dtype=float32>]
(Pdb) tf.gradients(t, self._input_values)
[<tf.Tensor 'gradients_1/layer_0/MatMul_grad/MatMul:0' shape=(?, 20) dtype=float32>]
```

Reply:
Sorry, still a bit confused here. The docs say that gradients returns "A list of sum(dy/dx) for each x in xs." Is the sum "for each x in xs" over the columns in a given instance or over instances?

The first option (sum over columns) doesn't make sense to me because then you don't get per-column gradients, i.e. feature importances. The second option doesn't make sense with the indexing because it seems like `gradients[0]` would just give you dy/dx_0.

For the non-multiclass case I'm assuming:
```
t.shape = (n_examples,)
self._input_values.shape = (n_examples, n_features)

tf.gradients().shape = (n_examples,)  # sum over columns?
tf.gradients().shape = (n_features,)  # sum over instances?
```
Where am I going wrong here?

Reply:
Perhaps this will help (sorry this is tricky and I didn't explain it well)

```python
# Here's an illustration of tf.gradients behavior for a linear transformation.
import numpy as np
import tensorflow as tf
# n-by-3
x = tf.placeholder(dtype=tf.float32, shape=(None, 3))
# 3-by-2
w_np = np.array([[1, 2],
                 [3, 4],
                 [5, 6]], dtype=np.float32)
w = tf.constant(w_np)
# n-by-2
y = tf.matmul(x, w)
print("output shape:", y.get_shape())
# n-by-3
g_multioutput = tf.gradients(y, x)
g_singleoutput = tf.gradients(y[:,0], x)
session = tf.Session()
x_np = np.zeros((10, 3), dtype=np.float32)
result_singleoutput = session.run(g_singleoutput, feed_dict={x: x_np})
print("dy[0]/dx:\n", result_singleoutput, sep="")
result_multioutput = session.run(g_multioutput, feed_dict={x: x_np})
print("dy/dx:\n", result_multioutput, sep="")
# The gradient is the same for each row of input because this is just linear.
# The gradient will just be the column of the linear transformation w,
# or the sum across columns for the multioutput case since tf.gradients
# will sum across the columns of y.
for row_single, row_multi in zip(result_singleoutput, result_multioutput):
    assert np.allclose(row_single, w_np[:, 0])
    assert np.allclose(row_multi, w_np.sum(axis=1))
```

Reply:
Ok I think I get this now - the `[0]` is because we get back a length - 1 list, but the element in that list is the array of n_examples x n_features with the gradients for each feature for each instance. The `tf.stack` business is because we don't want to sum over y dimensions in the multiclass case (in case a gradient is + for some dv but - for another).

I think my confusion was that you get back a n_features x n_gradients array wrapped in a list - I thought you would just get back the array, which is more of an issue with my reading of the tf docs. When I read "for each `x` in `xs`" my first interpretation is to think of the single `x` as a row, not as a tensor.